### PR TITLE
16 get product by

### LINKED ISSUE #16: Back-end Route| GET |/product/:productId
#### product by ID Route
gets the data of the product for this ID 
#### Protected: False

---

#### Main Things : 
- [ ] Database
- [ ] Response
- [ ] Test

---

#### Database :
Use database query to get order data from the database 

---
#### Response : 
Response may be success or error :
* Error :
```javascript
    {
        status:400,
        messgae:'Bad reqeust'
    }
```
* Success
```javascript
    {
        status:200,
    data:[{
        id:1,
        name : "حذاء" ,
        description : " حذاء ابيض ماركة تومي " ,
        img : [{ URL : "https://dsgfhgjhkjlk"} , {...} , ....],
        category : "رياضة",
        rating : 10,
        old_price : 6 ,
        new_price : 5 ,
        quantity : 5,
        add_date : 22/6/2020,
        
        }]
    }
```
#### Test : 
* Route : 
  Test route by testing all the expected cases for correct retrieve for data from the products table.
  
* Query : 
  Test select database query to make sure that everything is ok.

### PR DESCRIPTION
**Pull request by AlaaShurrab**
fixes #16 
- creating a database query to fetch the data of one product using the params as shown in the issue 
- adding the route and connecting the route to the API 
- testing for the database query and for the route

### CODE REVIEW

**Review by Sondos-Alalami — DISMISSED** [review]

**Review by Sondos-Alalami — DISMISSED** [review]

File: __test__/getProductById.test.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+const request = require('supertest');
+const app = require('../server/app');
```
Comment (mFlaifel) [inline __test__/getProductById.test.js:3]: please add an empty line here to separate your modules from the third-party modules

File: __test__/getProductById.test.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+const connection = require('../server/database/connection');
+const buildDb = require('../server/database/data/build');
+const { getProductById } = require('../server/database/queries');
+
+describe('Test Routes, Test Database getAllProducts query', () => {
```
Comment (mFlaifel) [inline __test__/getProductById.test.js]: please change the description 
```suggestion
describe(`Test '/product/:productId' route, and test  get product by id  query`, () => {
```

File: server/controllers/common/productByIdHandler.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const { getProductById } = require('../../database/queries');
+
+const productByIdHandler = (req, res, next) => {
+  getProductById(req.params.productId)
+    .then(({ rows }) => res.status(200).json({
+      status: 200,
+      data: rows,
+    }))
+    .catch(next);
+};
```
Comment (mFlaifel) [inline server/controllers/common/productByIdHandler.js]: if the product id doesn't exist, send a message like ' product does not exist'
```
res.json({
status:404,
message: 'product does not exist',
data: [ ]
})
```

File: server/database/queries/index.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
 const getAllProducts = require('./getAllProducts');
 const getClients = require('./getClients');
+const getProductById = require('./getProductById');
 const getOrders = require('./getOrders');
```
Comment (mFlaifel) [inline server/database/queries/index.js]: please arrange all related routes by their category  (common routes  first , then buyer and finally seller)
```suggestion
const getAllProducts = require('./getAllProducts');
const getProductById = require('./getProductById');
const getClients = require('./getClients');
const getOrders = require('./getOrders');
```

File: server/database/queries/index.js [hunk 5]
```diff
@@ -1,9 +1,11 @@
...
 
 module.exports = {
   getClients,
   getOrders,
   getAllProducts,
+  getProductById,
```
Comment (mFlaifel) [inline server/database/queries/index.js]: (common routes  first , then buyer and finally seller)
```suggestion
  getAllProducts,
  getProductById,
  getClients,
  getOrders,
```

**Review by mFlaifel — CHANGES_REQUESTED** [review]

File: __test__/getProductById.test.js (hunk 1 above)
Comment (AlaaShurrab) [reply __test__/getProductById.test.js:3]: done

File: __test__/getProductById.test.js (hunk 2 above)
Comment (AlaaShurrab) [reply __test__/getProductById.test.js]: done

File: server/controllers/common/productByIdHandler.js (hunk 3 above)
Comment (AlaaShurrab) [reply server/controllers/common/productByIdHandler.js]: done, also added a test to check this case out

**Review by mFlaifel — APPROVED** [review]

**Pull request merged**